### ISSUE TITLE
Set dialog position depending on mouse position

### ISSUE THREAD

**Issue by Seiifer** [likeastore/ngDialog#222]
Hi, thanks for this usefull tool you are providing.

I was wondering :

Is there a way to set the position of a new dialog ? 
I tried a few things with appendTo but it didn't work for me. I wish to open mydialog depending on the current position of the mouse. The ideal would be to :
- set the up-left dialog's corner position  to the current mouse position if the mouse is closer to the up-left corner of the window.
- set the up-right dialog's corner position  to the current mouse position if the mouse is closer to the up-right corner of the window.
- set the down-left dialog's corner position  to the current mouse position if the mouse is closer to the down-left corner of the window.
- set the down-right dialog's corner position  to the current mouse position if the mouse is closer to the down-right corner of the window.

Also, is there a way to specify a dialog's width and height ? And making its content scrollable if it requires more than the given height (scrollable or height-extendable by clicking on a specific element in the bottom of the dialog otherwise)

Regards

**Comment by egor-smirnov** [comment]
@Seiifer 

About the first part with positioning. I won't introduce something related to mouse position into API of the ngDialog. I would better introduce some `position` option that let developer override the position where dialog will appear. In this case, you as a user of the plugin will decide what are the coordinates (you will set them according to your algorithm). You thoughts?

About second part with width / height / scrolling. It's more CSS related. Take a look at #251

**Comment by Seiifer** [comment]
@egor-smirnov 

Adding a position parameter looks great and will suit most of user cases, but to be honest I was wondering if there wasn't already something allowing it since there is some append function.

Regarding the second part, I'm already using the plain theme with custom width that allows me to choose the width of the dialog.
But I don't know how to specify a maximum height for the dialog, allowing us to activate scrolling in the dialog if its contents takes more than its height.

**Comment by egor-smirnov** [comment]
@Seiifer so I am going to add positioning support for the next versions. Thanks for idea.

Regards max height - will `height:..px;max-height:..px;overflow:auto;` work for you?

**Comment by Seiifer** [comment]
Regarding the height, you intend to make the dialog content scrollable if it reaches the max height ?

Ideally, the best would be to be able to set width and height. For example, replacing `custom-width` with `custom-dimensions`  then we just need to change the content height and width in css.

**Comment by egor-smirnov** [comment]
> Regarding the height, you intend to make the dialog content scrollable if it reaches the max height ?

Yes. About height things - please create separate issue or submit PR with more precise explanations. 

I am going to keep this issue for tracking progress related to positioning of the dialog.

**Comment by bfree-github** [comment]
@egor-smirnov

I need to change the position of the dialog for small devices (via @media).  I've been looking for a ngDialog position parameter, but couldn't find anything... was about to go through the code to add one myself, when I saw this posting...

Any idea when you'll have this change posted?

Alternately, if the position could be relative to the display area (as opposed to the page size), then I could avoid @media altogether.

Much thanks - awesome job!

**Comment by egor-smirnov** [comment]
@bfree-github looking at doing this in upcoming 1-2 weeks, but can't be precise; quite busy at my job last days.

You could submit PR if you want this ASAP.

**Comment by bfree-github** [comment]
Understood - thanks!

**Comment by bfree-github** [comment]
Oh wow - going through the code, I found that padding-top does exactly what I'm looking for.

I just set it to 15% and it works the way I want on all devices!

Thanks!

**Comment by gottfrois** [comment]
A `position` option would be great

**Comment by NicolaiKilian** [comment]
Hello,

I've written a couple of services to make the dialogs adapt automatically to the window height:

First, send an event whenever the window is resized:
```javascript
angular.module('myApp').directive('resize', function($window, dimensionService) {
    return {
        link: function(scope, element, attrs) {
            angular.element($window).on('resize', function() {
                scope.$broadcast('resize::resize', dimensionService.getWindowDimension());
            });
        }
    };
});
```
```javascript
angular.module('myApp').factory('dimensionService', function (ngDialog, $window, $document) {
    var DimensionService = {};

    DimensionService.getWindowDimension = function() {
        var w = $window,
            d = $document[0],
            e = d.documentElement,
            g = d.getElementsByTagName('body')[0],
            x = w.innerWidth || e.clientWidth || g.clientWidth,
            y = w.innerHeight|| e.clientHeight|| g.clientHeight;

        return {'innerWidth' : x, 'innerHeight' : y};
    };

    return DimensionService;
});
```

```html
<body ng-app="myApp" resize>
```

Then, listen to both the resize event and the dialog-loaded event and set the dialog content panel's height accordingly:

```javascript
angular.module('myApp').factory('dialogService', function (ngDialog, dimensionService) {
    var DialogService = {};

    DialogService.open = function(scope, dialogParams) {
        var dialog = ngDialog.open(dialogParams);

        function updateMaxHeight(windowInnerHeight) {
            $('#' + dialog.id + ' .ngdialog-content').attr('style', 'max-height:' + (windowInnerHeight * 0.8) + 'px; overflow-y: auto');
        }

        scope.$on('ngDialog.opened', function (event, $dialog) {
            updateMaxHeight(dimensionService.getWindowDimension().innerHeight);
        });

        scope.$on('resize::resize', function(event, windowSize) {
            updateMaxHeight(windowSize.innerHeight);
        });
    };

    return DialogService;
});
```

Hope that helps you!

**Comment by bfree-github** [comment]
That's great - much thanks!

**Comment by NicolaiKilian** [comment]
you're welcome! Change the factor 0.8 as needed

**Comment by f1ght4fun** [comment]
Hello,

Is there any updates on adding 'position' to available dialog options?

Tnx

**Comment by egor-smirnov** [comment]
Currently no, I am too busy at my full-time job. You could contribute by writing actual code & opening new PR. This will be appreciated a lot.

This feature request is on top of priority list and I am going to. But feel free to contribute if you need it ASAP. Just warn me beforehand if any of you is going to implement this yourself in the nearest time. This is needed not to do the same work later from my side.

Thanks for understanding.

**Comment by bfree-github** [comment]
I might have some bandwidth to work on this next week.  I'll let you know what how I plan to do it before submitting changes.  Thanks!

**Comment by egor-smirnov** [comment]
@bfree-github :+1:

**Comment by faceleg** [comment]
@bfree-github bump

**Comment by iambkg** [comment]
Hi there!
Is there any news about position attribute? Using your lib and need this option so much...
Thanks in advance!

**Comment by egor-smirnov** [comment]
@iambkg no news from our side :( Until the next year I won't have enough time for ngDialog.
@bfree-github maybe you have some time?

**Comment by bfree-github** [comment]
Sorry for my delay - I just came out of ER for angioplasty and 2 stents for 3 blocked arteries.

I’ll try to look into after I recover - Bob

**Comment by iambkg** [comment]
Hi guys!
@egor-smirnov, @bfree-github thank you for your responses, we have a workaround - position dialogs via CSS. So, positioning via binding is not so major issue.
Thank you all for your work!
Regards,
Alexey

**Comment by faceleg** [comment]
Hope you recover swiftly and fully, @bfree-github

**Comment by markzolotoy** [comment]
@iambkg
Can you please share with us how exactly you are positioning the dialog window?

**Comment by iambkg** [comment]
@markzolotoy 
Sure. (Sorry for delay).
We use positioning via css, set it from code. May it's not the best way, but this approach works for us.
(We use TypeScript)
````typescript
let dialogContainer: JQuery = $('.dialogClassName'); // get dialog container by class name
if (dialogContainer.length > 0) {
    const dialogTop: number = 15; // hardcode just for example, we use some calculations here
    const dialogLeft: number = 20; // here too
    dialogContainer.css({top: dialogTop + 'px', left: dialogLeft + 'px'});
}
````

**Comment by waterplea** [comment]
Any news on adding position parameter to the ngDialog? :)

**Comment by faceleg** [comment]
There will be no news until someone submits a PR :dancer:

**Comment by shannonlal** [comment]
I have recently used ngDialog on a project and was able to figure out how to move the placement of the dialog by changing the css (i.e. changing the margin for the .ngdialog-content).  I would be willing to implement a fix for this but wanted to check if the following solution would work:

1. Using an ng-click capture the event and determine the x and y position
2. Adjust the css (.ngdialog-theme-default .ngdialog-content)  for the margin so that the correspond to the  x and y position for the event. 

If this seems like a feasible solution or if you have alternative solutions please let me know.

Thanks

Shannon

**Comment by bhollis** [comment]
I might suggest using [jQuery UI position](https://jqueryui.com/position/) and accepting the same arguments - that would allow for nicely expressing the position of the dialog relative to another element, with flexibility around how it behaves around edges of the viewport.

**Comment by faceleg** [comment]
We're not adding a jQuery or jQuery ui dependency for this project

**Comment by bhollis** [comment]
the `position` module is super useful and easy to divorce from jQuery itself (it has no dependencies on the rest of jQuery UI either).
